lib/writers/s3.go

```diff
@@ -52,59 +52,119 @@ func NewS3Writer(logger logr.Logger, stateStoreSpec platformv1alpha1.BucketState
 	}, nil
 }
 
-func (b *S3Writer) WriteObject(objectName string, toWrite []byte) error {
+func (b *S3Writer) WriteDirWithObjects(deleteExistingContentsInDir bool, dir string, toWrite ...platformv1alpha1.Workload) error {
 	logger := b.Log.WithValues(
 		"bucketName", b.BucketName,
 		"path", b.path,
-		"objectName", objectName,
 	)
 
-	objectFullPath := filepath.Join(b.path, objectName)
-	if len(toWrite) == 0 {
-		logger.Info("Empty byte[]. Nothing to write to bucket")
-		return nil
-	}
-
+	objectsToDeleteMap := map[string]minio.ObjectInfo{}
 	ctx := context.Background()
 
-	// Check to see if we already own this bucket (which happens if you run this twice)
-	exists, errBucketExists := b.RepoClient.BucketExists(ctx, b.BucketName)
-	if errBucketExists != nil {
-		logger.Error(errBucketExists, "Could not verify bucket existence with provider")
-	} else if !exists {
-		logger.Info("Bucket provided does not exist (or the provided keys don't have permissions)")
+	//Get a list of all the old workload files, we delete any that aren't part of the new workload at the end of this function.
+	if deleteExistingContentsInDir {
+		var err error
+		objectsToDeleteMap, err = b.getObjectsInDir(ctx, dir, logger)
+		if err != nil {
+			return err
+		}
 	}
 
-	contentType := "text/x-yaml"
-	reader := bytes.NewReader(toWrite)
+	for _, item := range toWrite {
+		objectFullPath := filepath.Join(b.path, dir, item.Filepath)
+		logger := b.Log.WithValues(
+			"objectName", objectFullPath,
+		)
+		if len(toWrite) == 0 {
+			logger.Info("Empty byte[]. Nothing to write to bucket")
+			return nil
+		}
 
-	objStat, err := b.RepoClient.StatObject(ctx, b.BucketName, objectFullPath, minio.GetObjectOptions{})
-	if err != nil {
-		if minio.ToErrorResponse(err).Code == "NoSuchKey" {
-			logger.Info("Object does not exist yet")
+		ctx := context.Background()
+
+		// Check to see if we already own this bucket (which happens if you run this twice)
+		exists, errBucketExists := b.RepoClient.BucketExists(ctx, b.BucketName)
+		if errBucketExists != nil {
+			logger.Error(errBucketExists, "Could not verify bucket existence with provider")
+		} else if !exists {
+			logger.Info("Bucket provided does not exist (or the provided keys don't have permissions)")
+		}
```

Comment:
refactor: do we need to test this for each file? or just once before the loop?

Reply:
good spot, moved outside